plugins.d/loopsleepms.sh.inc

```diff
@@ -31,6 +22,92 @@ current_time_ms() {
     fi
 }
 
+# -----------------------------------------------------------------------------
+# use /proc/uptime as a high resolution timer
+
+current_time_ms_from_date
+current_time_ms_from_uptime_started="${now_ms}"
+current_time_ms_from_uptime_first=0
+current_time_ms_from_uptime() {
+    local up rest arr n
+
+    read up rest </proc/uptime
+    if [ $? -ne 0 ]
+        then
+        echo >&2 "$0: Cannot read /proc/uptime - falling back to current_time_ms_from_date()."
+        current_time_ms="current_time_ms_from_date"
+        current_time_ms_from_date
+        current_time_ms_accuracy=1
+        return
+    fi
+
+    arr=(${up//./ })
+
+    if [ ${#arr[1]} -lt 1 ]
+        then
+        n="${arr[0]}000"
+    elif [ ${#arr[1]} -lt 2 ]
+        then
+        n="${arr[0]}${arr[1]}00"
+    elif [ ${#arr[1]} -lt 3 ]
+        then
+        n="${arr[0]}${arr[1]}0"
+    else
+        n="${arr[0]}${arr[1]}"
+    fi
+
+    now_ms=$((current_time_ms_from_uptime_started - current_time_ms_from_uptime_first + n))
+
+    if [ "${now_ms}" -lt "${current_time_ms_from_uptime_started}" ]
+        then
+        echo >&2 "$0: Cannot use current_time_ms_from_uptime() - falling back to current_time_ms_from_date()."
+        current_time_ms="current_time_ms_from_date"
+        current_time_ms_from_date
+        current_time_ms_accuracy=1
+    fi
+}
+current_time_ms_from_uptime
+current_time_ms_from_uptime_first="$((now_ms - current_time_ms_from_uptime_started))"
+current_time_ms="current_time_ms_from_uptime"
+current_time_ms_accuracy=10
+if [ "${current_time_ms_from_uptime_first}" -eq 0 ]
+    then
+    echo >&2 "$0: Invalid setup for current_time_ms_from_uptime() - falling back to current_time_ms_from_date()."
+    current_time_ms="current_time_ms_from_date"
+    current_time_ms_accuracy=1
+fi
+
+# -----------------------------------------------------------------------------
+# use read with timeout for sleep
+
+mysleep="mysleep_read"
+
+mysleep_fifo="${NETDATA_CACHE_DIR-/tmp}/.netdata_bash_sleep_timer_fifo"
+[ -e "${mysleep_fifo}" ] && rm "${mysleep_fifo}"
+mkfifo "${mysleep_fifo}" || mysleep="sleep"
+
+mysleep_read() {
+    read -t "${1}" <>"${mysleep_fifo}"
+    ret=$?
+    if [ $ret -le 128 ]
+        then
+        echo >&2 "$0: Cannot use read for sleeping (return code ${ret})."
+        mysleep="sleep"
+        ${mysleep} "${1}"
+    fi
+}
+
```

Comment:
the function above sleeps for the given decimal amount of seconds, using `read` and a fifo.